### PR TITLE
peer: Brontide Peer implementation

### PR DESCRIPTION
**Pull request by Crypt-iQ**
Creates `peer` package. Incrementally moves code to the peer package, improves code comments and readability. Also fixes peer tests that weren't checking for a broadcasted transaction. `server`, `rpcserver` now both point to the `peer` package.

Closes https://github.com/lightningnetwork/lnd/issues/4186

### CODE REVIEW

**Comment by Crypt-iQ** [comment]
Integration tests currently failing on the new itest error whitelist. Not sure if there is an error, or if I need to add something to the whitelist?

File: peer.go [hunk 1]
```diff
@@ -624,12 +618,12 @@ func (p *peer) loadActiveChannels(chans []*channeldb.OpenChannel) (
 	return msgs, nil
 }
 
-// addLink creates and adds a new link from the specified channel.
+// addLink creates and adds a new ChannelLink from the specified channel.
 func (p *peer) addLink(chanPoint *wire.OutPoint,
 	lnChan *lnwallet.LightningChannel,
 	forwardingPolicy *htlcswitch.ForwardingPolicy,
 	chainEvents *contractcourt.ChainEventSubscription,
-	currentHeight int32, syncStates bool) error {
```
Comment (cfromknecht) [inline peer.go:632]: `currentHeight` was unused?? good to remove then 👍

**Comment by cfromknecht** [comment]
@Crypt-iQ linter is failing due to recently merged prs in master

File: peer/config.go [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+	// write buffers allocated at any one time, and decouple write buffer
+	// allocation from the peer life cycle.
+	WritePool *pool.Write
+
+	// ReadPool is the task pool that manages reuse of read buffers.
+	ReadPool *pool.Read
+
+	// Switch is a pointer to the htlcswitch. It is used to setup, get, and
+	// tear-down ChannelLinks.
+	Switch *htlcswitch.Switch
```
Comment (Roasbeef) [inline peer/config.go:86]: In the future, we can abstract this with a _slimmer_ interface. Same goes for the other raw pointers below.

File: peer/config.go [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+	// InterceptSwitch is a pointer to the InterceptableSwitch, a wrapper around
+	// the regular Switch. We only export it here to pass ForwardPackets to the
+	// ChannelLinkConfig.
+	InterceptSwitch *htlcswitch.InterceptableSwitch
```
Comment (Roasbeef) [inline peer/config.go:91]: Hmm, don't we only need this one given that it embeds the above?

File: peer/config.go [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+	// PrunePersistentPeerConnection is used to remove all internal state
+	// related to this peer in the server.
+	PrunePersistentPeerConnection func([33]byte)
+
+	// FetchLastChanUpdate fetches our latest channel update for a target
+	// channel.
+	FetchLastChanUpdate func(lnwire.ShortChannelID) (*lnwire.ChannelUpdate,
+		error)
+
+	// ProcessFundingOpen is used to hand off an OpenChannel message to the
```
Comment (Roasbeef) [inline peer/config.go:175]: Non-blocking, but in the future we could break these up into embedded sub-configs to make this less of a "mega struct".

File: peer/config.go [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+	// funding manager.
+	ProcessFundingSigned func(*lnwire.FundingSigned, lnpeer.Peer)
+
+	// ProcessFundingLocked is used to hand off a FundingLocked message to the
+	// funding manager.
+	ProcessFundingLocked func(*lnwire.FundingLocked, lnpeer.Peer)
+
+	// ProcessFundingError is used to hand off an Error message to the funding
+	// manager.
+	ProcessFundingError func(*lnwire.Error, *btcec.PublicKey)
```
Comment (Roasbeef) [inline peer/config.go:197]: How about a handler for any unknown messages? Could eventually allow users to import `lnd` into another project then manually handle some new extension they're experimenting with.

File: peer/config.go [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+	// MaxChannelFeeAllocation is used when creating ChannelLinks and is the
+	// maximum percentage of total funds that can be allocated to a channel's
+	// commitment fee. This only applies for the initiator of the channel.
+	MaxChannelFeeAllocation float64
+
+	// ServerPubKey is the serialized, compressed public key of our lnd node.
+	// It is used to determine which policy (channel edge) to pass to the
+	// ChannelLink.
+	ServerPubKey []byte
```
Comment (Roasbeef) [inline peer/config.go:224]: `[33]byte`?

File: server.go [hunk 7]
```diff
@@ -1480,7 +1480,13 @@ func (s *server) Stop() error {
 		// Disconnect from each active peers to ensure that
 		// peerTerminationWatchers signal completion to each peer.
 		for _, peer := range s.Peers() {
-			s.DisconnectPeer(peer.addr.IdentityKey)
+			err := s.DisconnectPeer(peer.IdentityKey())
+			if err != nil {
+				srvrLog.Errorf("could not disconnect peer: %v"+
```
Comment (Roasbeef) [inline server.go:1478]: `Warnf` instead? Otherwise may trigger some false positives in the post-submit error scanning.

File: peer.go [hunk 8]
```diff
@@ -449,11 +449,10 @@ func (p *peer) Start() error {
 // initGossipSync initializes either a gossip syncer or an initial routing
 // dump, depending on the negotiated synchronization method.
 func (p *peer) initGossipSync() {
-	switch {
 
 	// If the remote peer knows of the new gossip queries feature, then
 	// we'll create a new gossipSyncer in the AuthenticatedGossiper for it.
-	case p.remoteFeatures.HasFeature(lnwire.GossipQueriesOptional):
+	if p.remoteFeatures.HasFeature(lnwire.GossipQueriesOptional) {
```
Comment (Roasbeef) [inline peer.go:377]: M-m-muh switch statements! 

Solid change though, can be upgraded back once we actually have additional cases.

File: peer/brontide.go [hunk 9]
```diff
@@ -104,13 +112,13 @@ type TimestampedError struct {
...
 
-// peer is an active peer on the Lightning Network. This struct is responsible
+// Brontide is an active peer on the Lightning Network. This struct is responsible
 // for managing any channel state related to this peer. To do so, it has
 // several helper goroutines to handle events such as HTLC timeouts, new
 // funding workflow, and detecting an uncooperative closure of any active
 // channels.
 // TODO(roasbeef): proper reconnection logic
-type peer struct {
+type Brontide struct {
```
Comment (Roasbeef) [inline peer/brontide.go:121]: I think this name works mainly to avoid having `peer.Peer` everywhere, but once this is merged we could go with another name since this isn't _strictly_ tied to brontide, though it does assert that specific connection from a regular `net.Conn`.

File: server.go (hunk 7 above)
Comment (Roasbeef) [reply server.go:1478]: I'd consider this the only blocking statement in this review.

**Comment by cfromknecht** [comment]
@Crypt-iQ linter also appears to be failing on some `goimports` check

File: peer/config.go (hunk 3 above)
Comment (Crypt-iQ) [reply peer/config.go:91]: From what I can tell it doesn't embed it, we'd need to either embed it or make a getter

File: peer/config.go (hunk 4 above)
Comment (Crypt-iQ) [reply peer/config.go:175]: Yeah I was thinking a FundingManager interface could be exposed in a later PR?

File: peer/brontide.go (hunk 9 above)
Comment (Crypt-iQ) [reply peer/brontide.go:121]: Yeah I wasn't sure about the name, also there's an `lnpeer.Peer` interface that it implements. I thought it would get too wordy.

File: server.go (hunk 7 above)
Comment (Crypt-iQ) [reply server.go:1478]: 👍

File: peer/config.go (hunk 6 above)
Comment (Crypt-iQ) [reply peer/config.go:224]: 👍

File: peer/config.go (hunk 5 above)
Comment (Crypt-iQ) [reply peer/config.go:197]: I think this should be deferred to a followup PR, this one was mostly meant to be a code move. I agree that there should be a handler for unknown messages though.

File: peer/brontide.go (hunk 9 above)
Comment (cfromknecht) [reply peer/brontide.go:121]: Yeah I think it’s fine to drop `net.Conn` and focus only on `*brontide.Conn`, you can’t really use anything else because the framing semantics are different from regular connections

File: peer/config.go (hunk 3 above)
Comment (Crypt-iQ) [reply peer/config.go:91]: But doing so is out of scope here, can do in a follow up when we slim down these raw pointers

File: peer/brontide.go (hunk 9 above)
Comment (Crypt-iQ) [reply peer/brontide.go:121]: Does having a `net.Conn` make testing any easier? That's the *only* reason I considered keeping it around. But if all MUST be *brontide.Conn what's the point

File: peer_test.go [hunk 10]
```diff
@@ -290,7 +290,7 @@ func TestPeerChannelClosureFeeNegotiationsResponder(t *testing.T) {
 	}
 
 	// Bob doesn't agree with the fee and will send one back that's 2.5x.
-	preferredRespFee := aliceClosingSigned.FeeSatoshis
+	preferredRespFee := responderClosingSigned.FeeSatoshis
```
Comment (cfromknecht) [inline peer_test.go]: needs squash

File: server.go [hunk 11]
```diff
@@ -2797,15 +2798,62 @@ func (s *server) peerConnected(conn net.Conn, connReq *connmgr.ConnReq,
...
-	p, err := newPeer(
-		s.cfg, conn, connReq, s, peerAddr, inbound, initFeatures,
-		legacyFeatures, s.cfg.ChanEnableTimeout,
-		lncfg.DefaultOutgoingCltvRejectDelta, errBuffer,
-	)
-	if err != nil {
-		srvrLog.Errorf("unable to create peer %v", err)
-		return
-	}
+	pCfg := ppeer.Config{
```
Comment (cfromknecht) [inline server.go]: it's beautiful 😥

nit: `peerConfig`?

File: peer/brontide.go (hunk 9 above)
Comment (cfromknecht) [reply peer/brontide.go:121]: no not really, a `brontide.Conn` implements `net.Conn` in case you wanted to use it in a separate context. in the peer however we don't use `Read` or `Write` because the protocol operates at the _message_ to handle timeouts properly. so in reality we only use `ReadMessage`, `WriteMessage`, and `Flush`

File: server.go (hunk 11 above)
Comment (Crypt-iQ) [reply server.go]: Will have to move some lines around if we do this (the copying of public keys right before the call to `newPeer` / `NewBrontide` depending on the commit. Is it worth it?

Comment (Crypt-iQ) [reply server.go]: Eh I don't mind the way it is, plus would have to move some lines around bc we have to copy the keys and they'd be too long.

Comment (cfromknecht) [reply server.go]: sgtm
